src/square.py

```diff
@@ -78,20 +78,38 @@ class Square(Vector):
 
     notation_range = re.compile(f"[{file_range}][{rank_range}]")
 
+    def __new__(cls, square: str | Vector):
+        """Abort if attempted square is illegal.
+
+        This is used to make `Square` objects self-correcting upon behavior.
+        Failure of creation will be its own validity check.
+
+        Args:
+            square: A square in notation form or a vector (assuming the board origin is top-left in the latter case).
+
+        Returns:
+            `None` if square specification is illegal.
+        """
+        try:
+            if isinstance(square, str):
+                assert Square.notation_range.match(square)
+            if isinstance(square, Vector):
+                assert 0 <= square.rank < 8 and 0 <= square.file < 8
+            return super(Square, cls).__new__(cls)
+        except AssertionError:
+            return None
+
```

Comment:
`Square` objects are now always self-constrained within proper chessboard boundaries, discarding a trivial high level check for legal moves.